lib/versioning/regex/index.ts

```diff
@@ -0,0 +1,151 @@
+const semver = require('semver');
+const { VersioningApi, RangeStrategy } = require('../common');
+const { logger } = require('../../logger');
+
+interface Version {
+  // major, minor, and patch act in the standard semver fashion, but without
+  // correctness requirements: if any one or two is omitted, they are treated
+  // as zero. At least one must be present
+  major: number;
+  minor: number;
+  patch: number;
+  // prereleases are treated in the standard semver manner, if present
+  prerelease: string;
+  // architectures, if present, are treated as a compatibility layer: we will
+  // never try to update to a version with a different architecture. Other than
+  // for juding compatibility (exact string match), the architecture is ignored
+  architecture: string;
+}
+
+// convenience method for passing a Version object into any semver.* method.
+function asSemver(version: Version): string {
+  let vstring = `${version.major}.${version.minor}.${version.patch}`;
+  if (typeof version.prerelease !== 'undefined') {
+    vstring += `-${version.prerelease}`;
+  }
+  return vstring;
+}
+
+export const api: VersioningApi = {
+  // config is expected to be overridden by a user-specified RegExp value
+  // sample values:
+  //
+  // * emulates the "semver" configuration:
+  //   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<prerelease>.*))?$')
+  // * emulates the "docker" configuration:
+  //   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<architecture>.*))?$')
+  // * matches the versoining scheme used by the Python images on DockerHub:
+  //   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(?<prerelease>[^.-]+)?(-(?<architecture>.*))?$');
+  //
+  // defaults to a simple semver scheme
+  config: null,
+
+  // TODO: the linter expects method shorthands for all these methods, but
+  // those don't seem to bind `this`. What's the middle-ground?
+  configure: function(config) {
+    // without at least one of {major, minor, patch} specified in the regex,
+    // this versioner will not work properly
+    // TODO: there's got to be a better way to do this
+    if (config.includes('<major>') + config.includes('<minor>') + config.includes('<patch>') === 0) {
+      logger.fatal({ config }, 'Invalid versionConfig for versionScheme regex');
+      return;
+    }
+
+    // TODO: should we validate the user has not added extra unsupported
+    // capture groups?
+
+    this.config = RegExp(config);
+  },
+
+  parse: function(version: string): Version {
+    const match = version.match(this.config);
+    if (match === null) {
+      // TODO: what are we expected to return for invalid functions?
```

Comment:
@ contributors: how should the various `VersioningApi` methods return if a version returned as a possible update is completely invalid (say, if someone releases a version `foo` to an otherwise-semver package)? I assume `isValid() == false` would be enough here, but which of the `VersioningApi` methods do I need to implement that for? Can I assume if `isValid()` is false then none of the other methods will get called?

Reply:
Here the different version apis are not consistent. Some return `null` or `undefined` as `false`.
This should be fixed in a future version.